odl/tomo/backends/astra_cpu.py

```diff
@@ -0,0 +1,192 @@
+                                           astra_cleanup)
+from odl.tomo.geometry.geometry import Geometry
+from odl.tomo.geometry.fanbeam import FanFlatGeometry
+
+__all__ = ('astra_cpu_forward_projector_call',
+           'astra_cpu_backward_projector_call',
+           'ASTRA_AVAILABLE')
+
+
+# TODO: Fix inconsistent scaling of ASTRA projector with pixel size
+
+def astra_cpu_forward_projector_call(vol_data, geometry, proj_space, out=None):
+    """Run an ASTRA forward projection on the given data using the CPU.
+
+    Parameters
+    ----------
+    vol_data : `DiscreteLpVector`
+        Volume data to which the forward projector is applied
+    geometry : `Geometry`
+        Geometry defining the tomographic setup
+    proj_space : `DiscreteLp`
+        Space to which the calling operator maps
+    out : `DiscreteLpVector` or `None`, optional
+        Vector in the projection space to which the result is written.
+        Default: `None`
+
+    Returns
+    -------
+    out : ``proj_space`` element
+        Projection data resulting from the application of the projector
+    """
+    if not isinstance(vol_data, DiscreteLpVector):
+        raise TypeError('volume data {!r} is not a `DiscreteLpVector` '
+                        'instance.'.format(vol_data))
+    if not isinstance(geometry, Geometry):
+        raise TypeError('geometry  {!r} is not a `Geometry` instance.'
+                        ''.format(geometry))
+    if not isinstance(proj_space, DiscreteLp):
+        raise TypeError('projection space {!r} is not a `DiscreteLp` '
+                        'instance.'.format(proj_space))
+
+    if vol_data.space.grid.ndim != geometry.ndim:
+        raise ValueError('dimensions {} of volume data and {} of geometry '
+                         'do not match.'
+                         ''.format(vol_data.space.grid.ndim, geometry.ndim))
+
+    ndim = vol_data.space.grid.ndim
+
+    # Create astra geometries
+    vol_geom = astra_volume_geometry(vol_data.space)
+    # proj_geom = astra_projection_geometry(geometry, vol_data.space)
+    proj_geom = astra_projection_geometry(geometry)
+
+    if out is None:
+        out = proj_space.element()
+
+    # Create ASTRA data structures
+    vol_id = astra_data(vol_geom, datatype='volume', data=vol_data)
+    sino_id = astra_data(proj_geom, datatype='projection', data=out,
+                         ndim=proj_space.grid.ndim)
+
+    # Create projector
+    vol_interp = vol_data.space.interp
+    proj_id = astra_projector(vol_interp, vol_geom, proj_geom, ndim,
+                              impl='cpu')
+
+    # Create algorithm
+    algo_id = astra_algorithm('forward', ndim, vol_id, sino_id, proj_id,
+                              impl='cpu')
+
+    # Run algorithm and delete it
+    astra.algorithm.run(algo_id)
+
+    # Flip detector pixels for fanflat
+    if isinstance(geometry, FanFlatGeometry):
+        out[:] = out.asarray()[::-1, ::-1]
+
+    # Delete ASTRA objects
+    astra_cleanup()
```

Comment:
That was my dirty first implementation. The proper way of handling this would be to initialize a dictionary differentiating between data structure types, recording the IDs. The cleanup function should then (optionally) take this dictionary and free the space of those objects which are mentioned.